src/GraphQL/Utilities/Visitors/Custom/PatternMatchingVisitor.cs

```diff
@@ -0,0 +1,53 @@
+using System.Text.RegularExpressions;
+using GraphQL.Types;
+
+namespace GraphQL.Utilities.Visitors;
+
+/// <summary>
+/// A schema visitor that adds validation for the <see cref="PatternMatchingDirective"/> directive.
+/// Use as follows to enable validation of arguments and input fields:
+/// <code>
+/// services.AddGraphQL(b => b
+///     .AddSchema&lt;MySchema&gt;()
+///     .ConfigureSchema(s =>
+///     {
+///         s.Directives.Register(new <see cref="PatternMatchingDirective"/>());
+///         s.RegisterVisitor(new <see cref="PatternMatchingVisitor"/>());
+///     }));
+///
+/// field.Field(x => x.Email)
+///     .ApplyDirective("pattern", "regex", @".+\@.+\..+");
+/// </code>
+/// </summary>
+public class PatternMatchingVisitor : BaseSchemaNodeVisitor
+{
+    /// <inheritdoc/>
+    public override void VisitInputObjectFieldDefinition(FieldType field, IInputObjectGraphType type, ISchema schema)
+        => field.Validator += GetValidator(field);
+
+    /// <inheritdoc/>
+    public override void VisitObjectFieldArgumentDefinition(QueryArgument argument, FieldType field, IObjectGraphType type, ISchema schema)
+        => argument.Validator += GetValidator(argument);
+
+    private static Action<object?>? GetValidator(IMetadataReader fieldArgOrInputField)
+    {
+        // look for @pattern directive applied to the field argument or input field, and if found, use Validate to set the validation method
+        var applied = fieldArgOrInputField.FindAppliedDirective("pattern");
+        if (applied?.FindArgument("regex")?.Value is not string regex)
+            return null;
+
+        // if GlobalSwitches.DynamicallyCompileToObject is true, then compile
+        // the regex (or pull from the static cache if already compiled)
+        // (note that GlobalSwitches.DynamicallyCompileToObject is false for AOT schemas and scoped schemas)
+        var regexObject = new Regex($"^{regex}$",
+            GlobalSwitches.DynamicallyCompileToObject ? RegexOptions.Compiled : RegexOptions.None);
```

Comment:
Is there any trick to make it work with [GeneratedRegexAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.generatedregexattribute?view=net-8.0) that will generate the regex implementation which will be used here?

Reply:
I don't think so.  An optimized implementation for code-first would be like:

```cs
    [GeneratedRegex(
        @"^(ht|f)tp(s?)\:\/\/[0-9a-zA-Z]([-.\w]*[0-9a-zA-Z])*(:(0-9)*)*(\/?)([a-zA-Z0-9\-\.\?\,\'\/\\\+&amp;%\$#_]*)?$", 
        RegexOptions.CultureInvariant, 
        matchTimeoutMilliseconds: 1000)]
    private static partial Regex MatchIfValidUrl();

//type def
Field(x => x.Url)
    .Validate(value =>
    {
        if (!MatchIfValidUrl.IsMatch((string)value))
            throw new "Invalid URL";
    });
```

Unless we add a dedicated attribute for regex (this PR doesn't), and if somehow we can chain source builders so our source builder notices the attribute and links into the regex source builder to generate the code for that regex .....

Reply:
Looks like no generic solution. Only manual validator configuration like in your example.